### PR TITLE
feat(modem): host test support of the latest ESP-IDF release (IDF-7828)

### PR DESCRIPTION
**Pull request by gabsuren**

### CODE REVIEW

File: components/esp_modem/test/host_test/main/test_modem.cpp [hunk 1]
```diff
@@ -6,9 +6,10 @@
 #define CATCH_CONFIG_MAIN // This tells the catch header to generate a main
 #include <memory>
 #include <future>
-#include "catch.hpp"
+#include <catch2/catch_test_macros.hpp>
 #include "cxx_include/esp_modem_api.hpp"
 #include "LoopbackTerm.h"
+#include <iostream>
```
Comment (david-cermak) [inline components/esp_modem/test/host_test/main/test_modem.cpp:12]: Is this needed?

File: components/esp_modem/test/host_test/CMakeLists.txt [hunk 2]
```diff
@@ -4,9 +4,10 @@ include($ENV{IDF_PATH}/tools/cmake/project.cmake)
 
 set(EXTRA_COMPONENT_DIRS    # Add esp_modem component and linux port components
         ../..
-        ../../port/linux)
+        ../../port/linux
+        "$ENV{IDF_PATH}/tools/mocks/freertos/")
```
Comment (david-cermak) [inline components/esp_modem/test/host_test/CMakeLists.txt:8]: Why do we need freertos mocks?

Comment (gabsuren) [reply components/esp_modem/test/host_test/CMakeLists.txt:8]: https://github.com/gabsuren/esp-protocols-1/actions/runs/7100388129/job/19326385718

**Comment by gabsuren** [comment]
> LGTM in general.
> 
> It only seems like the tests were not executed? https://github.com/espressif/esp-protocols/actions/runs/7131318373/job/19419621023?pr=454#step:6:386
> 
> compared to the master pipeline: https://github.com/espressif/esp-protocols/actions/runs/7112254051/job/19361852029#step:6:321
> 
> the new test has no output, is it because of the new catch?

Yes in new Catch2 when the result is being redirected with `-r` option in `junit.xml` file it doesn't show the output. 
I tried to run in without -r, and it showed the amount of tests `pass/fail`
https://github.com/gabsuren/esp-protocols-1/actions/runs/7474518733/job/20340845490#step:6:408

**Comment by david-cermak** [comment]
> Yes in new Catch2 when the result is being redirected with `-r` option in `junit.xml` file it doesn't show the output. I tried to run in without -r, and it showed the amount of tests `pass/fail` https://github.com/gabsuren/esp-protocols-1/actions/runs/7474518733/job/20340845490#step:6:408

LGTM if the output is captured in `junit.xml`, could you please post a link?

**Comment by gabsuren** [comment]
> > Yes in new Catch2 when the result is being redirected with `-r` option in `junit.xml` file it doesn't show the output. I tried to run in without -r, and it showed the amount of tests `pass/fail` https://github.com/gabsuren/esp-protocols-1/actions/runs/7474518733/job/20340845490#step:6:408
> 
> LGTM if the output is captured in `junit.xml`, could you please post a link?

https://github.com/espressif/esp-protocols/actions/runs/7666426164/job/20894215662#step:6:413

**Comment by gabsuren** [comment]
@david-cermak I added `--reporter JUnit::out=result-junit.xml --reporter console::out=-::colour-mode=ansi` flags in case if we want to see the output and although redirect it to the file

https://github.com/espressif/esp-protocols/pull/454/files#diff-0e1a184c6942ac42834330ce6f9261da0dbd49d411e5f7847a38ae90fd3f8958R66:~:text=./build/%24%7B%7Binputs.app_name%7D%7D.elf%20%2D%2Dreporter%20JUnit%3A%3Aout%3Dresult%2Djunit.xml%20%2D%2Dreporter%20console%3A%3Aout%3D%2D%3A%3Acolour%2Dmode%3Dansi

**Comment by euripedesrocha** [comment]
> @david-cermak I added `--reporter JUnit::out=result-junit.xml --reporter console::out=-::colour-mode=ansi` flags in case if we want to see the output and although redirect it to the file
> 
> https://github.com/espressif/esp-protocols/pull/454/files#diff-0e1a184c6942ac42834330ce6f9261da0dbd49d411e5f7847a38ae90fd3f8958R66:~:text=./build/%24%7B%7Binputs.app_name%7D%7D.elf%20%2D%2Dreporter%20JUnit%3A%3Aout%3Dresult%2Djunit.xml%20%2D%2Dreporter%20console%3A%3Aout%3D%2D%3A%3Acolour%2Dmode%3Dansi
> 
> The log is here https://github.com/espressif/esp-protocols/actions/runs/7666426164/job/20894215662#step:6:413

You can run Catch with the `-s` option to have the successful assertions to be shown, but it is pretty verbose. I saw that we have a publish results that takes from xml output, where can it be seen?

File: .github/workflows/modem__build-host-tests.yml [hunk 3]
```diff
@@ -27,13 +27,19 @@ jobs:
             example: "simple_cmux_client"
             warning: "Warning: The smallest app partition is nearly full"
 
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
     container: espressif/idf:${{ matrix.idf_ver }}
     steps:
-      - name: Checkout esp-protocols
+      - name: Check out code (v3)
```
Comment (euripedesrocha) [inline .github/workflows/modem__build-host-tests.yml:33]: Why this is needed?

Comment (gabsuren) [reply .github/workflows/modem__build-host-tests.yml:33]: In the older IDF releases, specifically versions earlier than 5.0, it was not possible to execute Github Action checkout@v4. This issue arose because it required Node 20, which, in turn, necessitated GLIBC_2.28. 
And, installing GLIBC_2.28 was incompatible with the older IDF releases (Ubuntu 18)

**Comment by gabsuren** [comment]
> > @david-cermak I added `--reporter JUnit::out=result-junit.xml --reporter console::out=-::colour-mode=ansi` flags in case if we want to see the output and although redirect it to the file
> > https://github.com/espressif/esp-protocols/pull/454/files#diff-0e1a184c6942ac42834330ce6f9261da0dbd49d411e5f7847a38ae90fd3f8958R66:~:text=./build/%24%7B%7Binputs.app_name%7D%7D.elf%20%2D%2Dreporter%20JUnit%3A%3Aout%3Dresult%2Djunit.xml%20%2D%2Dreporter%20console%3A%3Aout%3D%2D%3A%3Acolour%2Dmode%3Dansi
> > The log is here https://github.com/espressif/esp-protocols/actions/runs/7666426164/job/20894215662#step:6:413
> 
> You can run Catch with the `-s` option to have the successful assertions to be shown, but it is pretty verbose. I saw that we have a publish results that takes from xml output, where can it be seen?

https://github.com/espressif/esp-protocols/actions/runs/7726114463#summary-21061763285

File: components/esp_modem/examples/linux_modem/CMakeLists.txt [hunk 4]
```diff
@@ -2,7 +2,8 @@ cmake_minimum_required(VERSION 3.5)
 
 include($ENV{IDF_PATH}/tools/cmake/project.cmake)
 
-set(EXTRA_COMPONENT_DIRS ../../port/linux)
+set(EXTRA_COMPONENT_DIRS ../../port/linux
+            "$ENV{IDF_PATH}/tools/mocks/freertos/")
```
Comment (david-cermak) [inline components/esp_modem/examples/linux_modem/CMakeLists.txt:6]: Why is this needed? esp_modem should be freertos agnostic, it has an OS abstraction layer, implemented in  `port/linux` for linux.

File: .github/workflows/modem__build-host-tests.yml [hunk 5]
```diff
@@ -76,10 +82,25 @@ jobs:
...
         app_name: "host_modem_test"
         app_path: "esp-protocols/components/esp_modem/test/host_test"
         component_path: "esp-protocols/components/esp_modem"
         upload_artifacts: true
+        run_executable: true
+        run_coverage: true
         pre_run_script: "esp-protocols/components/esp_modem/test/host_test/env.sh"
         publish_unit_test_result: true
+
+  host_test_linux_modem:
```
Comment (david-cermak) [inline .github/workflows/modem__build-host-tests.yml:95]: this is not a host test and we're only building it
```suggestion
  build_linux_example:
```

File: .github/workflows/modem__build-host-tests.yml (hunk 3 above)
Comment (david-cermak) [reply .github/workflows/modem__build-host-tests.yml:33]: could you please add this note to the comment, so we know why we cannot use `checkout@v4`

**Review by david-cermak — COMMENTED** [review]
LGTM otherwise

File: components/esp_modem/examples/linux_modem/CMakeLists.txt (hunk 4 above)
Comment (gabsuren) [reply components/esp_modem/examples/linux_modem/CMakeLists.txt]: From the log I see it uses freertos
https://github.com/espressif/esp-protocols/actions/runs/7782922741/job/21220340525#:~:text=%2D%2D%20Components%3A%20esp_common%20esp_event_mock,idf/components/soc

Comment (gabsuren) [reply components/esp_modem/examples/linux_modem/CMakeLists.txt]: @david-cermak After a closer look, I found that the `FreeRTOS` is coming as a common component, 
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/build-system.html#common-component-requirements:~:text=is%3A%20cxx%2C%20newlib%2C-,freertos,-%2C%20esp_hw_support%2C%20heap%2C%20log

https://gitlab.espressif.cn:6688/espressif/esp-idf/-/merge_requests/27455/diffs#338032a49374fdebda46efbb8226c68650fbeee5:~:text=%23%20target.-,%23%20This%20test%20app%20doesn%27t%20require%20FreeRTOS%2C%20using%20mock%20instead,list(APPEND%20EXTRA_COMPONENT_DIRS%20%22%24ENV%7BIDF_PATH%7D/tools/mocks/freertos/%22),-project(nvs_host_test

And as the test doesn't requires freertos, it uses mock instead
https://gitlab.espressif.cn:6688/espressif/esp-idf/-/merge_requests/27455/diffs#338032a49374fdebda46efbb8226c68650fbeee5

Comment (david-cermak) [reply components/esp_modem/examples/linux_modem/CMakeLists.txt:6]: > And as the test doesn't requires freertos, it uses mock instead 

If it was for the test, I would be okay(-ish) with it, I think adding a comment that the test needs and uses pthreads, but common_components need the FreeRTOS would be probably fine.

But this is an example, use of the freertos mock is simply wrong here.

File: .github/workflows/run-host-tests.yml [hunk 6]
```diff
@@ -51,14 +57,24 @@ jobs:
...
             cp sdkconfig.ci.linux sdkconfig.defaults
             idf.py build
-            ./build/${{inputs.app_name}}.elf -r junit -o junit.xml
+            if [ "${{ inputs.run_executable}}" == "false" ]; then
+                echo "Executeable wasn't run"
+                exit 0
+            fi
+
+            if [ "${{ inputs.publish_unit_test_result }}" == "true" ]; then
+              ./build/${{inputs.app_name}}.elf --reporter JUnit::out=junit.xml --reporter console::out=-::colour-mode=ansi
```
Comment (david-cermak) [inline .github/workflows/run-host-tests.yml:66]: does it still make sense to add these arguments if we hardcode them in the `app_main()`?

how is it with other host tests? Do they use the same action?

File: .github/workflows/run-host-tests.yml [hunk 7]
```diff
@@ -51,14 +57,19 @@ jobs:
             # The sdkconfig.ci.linux specifies Linux as the build target with apropriate settings.
             cp sdkconfig.ci.linux sdkconfig.defaults
             idf.py build
-            ./build/${{inputs.app_name}}.elf -r junit -o junit.xml
```
Comment (david-cermak) [inline .github/workflows/run-host-tests.yml:54]: This is a reusable action, correct? Wouldn't we miss the junit outputs for the websocket and mdns host tests?

**Review by david-cermak — APPROVED** [review]
LGTM, just one last final question. It's okay to just run the related tests to see if it didn't break anything...

File: .github/workflows/run-host-tests.yml (hunk 7 above)
Comment (gabsuren) [reply .github/workflows/run-host-tests.yml:54]: @david-cermak good point. I will modify and fix it in another PR. I'll address and resolve it in a subsequent pull request. As today I noticed a problem with the package upgrade for the mDNS test, but I believe the second commit in this pull request will fix the issue. 
Thank you for your review!
